core/src/main/java/org/infinispan/container/offheap/OffHeapConcurrentMap.java

```diff
@@ -4,63 +4,189 @@
+ * <p>
+ * For reference here is a list of commonly used terms:
+ * <ul>
+ *    <li><code>bucket</code>: Can store multiple entries (normally via a forward only list)
+ *    <li><code>memory lookup</code>: Stores an array of buckets - used primarily to lookup the location a key would be
+ *    <li><code>lock region</code>: A subset of the buckets that is protected by the same lock. There are N number of regions equal
+ * to the number of locks and each is equivalent in size. This will become smaller or larger during a resize event.
+ * </ul>
+ * <p>
+ * This implementation provides constant-time performance for the basic
+ * operations ({@code get}, {@code put}, {@code remove} and {@code compute}), assuming the hash function
+ * disperses the elements properly among the buckets.  Iteration over
+ * collection views requires time proportional to the number of buckets plus its size (the number
+ * of key-value mappings).  Thus, it's very important not to set the initial
+ * capacity too high if iteration performance is important. This map always assumes a load factor of .75 that is
+ * not changeable.
+ * <p>
+ * A map must be started after creating to create the initial memory lookup, which is also store in the native heap.
+ * When the size of the map reaches the load factor, that is .75 times the capacity, the map will attempt to resize
+ * by increasing its internal memory lookup to have an array of buckets twice as big. Normal operations can still
+ * proceed during this, allowing for minimal downtime during a resize.
+ * <p>
+ * This map is created assuming some knowledge of expiration in the Infinispan system. Thus operations that do not
+ * expose this information via its APIs are not supported. These methods are {@code keySet}, {@code containsKey} and
+ * {@code containsValue}.
+ * <p>
+ * This map guarantees consistency under concurrent read ands writes through a {@link StripedLock} where each
+ * {@link java.util.concurrent.locks.ReadWriteLock} instance protects an equivalent region of buckets in the underlying
+ * memory lookup. Read operations, that is ones that only acquire the read lock for their specific lock region, are
+ * ({@code get} and {@code peek}). Iteration on a returned entrySet or value collection will acquire only a single
+ * read lock at a time while inspecting a given lock region for a valid value. Write operations, ones that acquire the
+ * write lock for the lock region, including ({@code put}, {@code remove}, {@code replace}, {@code compute}. A clear
+ * will acquire all write locks when invoked. This allows the clear to also resize the map down to the initial size.
+ * <p>
+ * When this map is constructed it is also possible to provide an {@link EntryListener} that is invoked when various
+ * operations are performed in the map. Note that the various modification callbacks <b>MUST</b> free the old address,
+ * or else a memory leak will occur. Please see the various methods for clarification on these methods.
```

Comment:
Unfortunately, the eviction codes need to do the deallocation while holding the lru lock. This is because a write must modify the lru linked list which can modify a value in another lock region.

Reply:
I didn't mean removing the entry from the map, just the `allocator.deallocate()` call

Reply:
Yeah, I thought about redoing it when I changed the listener, but I didn't want to futz with this as well. Definitely could do this in a separate JIRA though.